packages/ws/src/strategies/sharding/WorkerShardingStrategy.ts

```diff
@@ -0,0 +1,211 @@
+
+/**
+ * Options for a {@link WorkerShardingStrategy}
+ */
+export interface WorkerShardingStrategyOptions {
+	/**
+	 * Dictates how many shards should be spawned per worker thread.
+	 */
+	shardsPerWorker: number | 'all';
+}
+
+/**
+ * Strategy used to spawn threads in worker_threads
+ */
+export class WorkerShardingStrategy implements IShardingStrategy {
+	private readonly manager: WebSocketManager;
+	private readonly options: WorkerShardingStrategyOptions;
+
+	private workers: Worker[] = [];
+	private readonly workerByShardId = new Collection<number, Worker>();
+
+	private readonly connectPromises = new Collection<number, () => void>();
+	private readonly destroyPromises = new Collection<number, () => void>();
+
+	private readonly throttler: IdentifyThrottler;
+
+	public constructor(manager: WebSocketManager, options: WorkerShardingStrategyOptions) {
+		this.manager = manager;
+		this.throttler = new IdentifyThrottler(manager);
+		this.options = options;
+	}
+
+	public async spawn(shardIds: number[]) {
+		const shardsPerWorker = this.options.shardsPerWorker === 'all' ? shardIds.length : this.options.shardsPerWorker;
+
+		let shards = 0;
+		while (shards !== shardIds.length) {
+			const slice = shardIds.slice(shards, shardsPerWorker + shards);
+			const workerData: WorkerData = {
+				...(await managerToFetchingStrategyOptions(this.manager)),
+				shardIds: slice,
+			};
+
+			const worker = new Worker(join(__dirname, 'worker.js'), { workerData });
+			await once(worker, 'online');
+			worker
+				.on('error', (err) => {
+					throw err;
+				})
+				.on('messageerror', (err) => {
+					throw err;
+				})
+				// eslint-disable-next-line @typescript-eslint/no-misused-promises
+				.on('message', (payload: WorkerRecievePayload) => this.onMessage(worker, payload));
+
+			this.workers.push(worker);
+			for (const shardId of slice) {
+				this.workerByShardId.set(shardId, worker);
+			}
+
+			shards += slice.length;
+		}
+	}
+
+	public async connect() {
+		const promises = [];
+
+		for (const [shardId, worker] of this.workerByShardId.entries()) {
+			await this.throttler.waitForIdentify();
+
+			const payload: WorkerSendPayload = {
+				op: WorkerSendPayloadOp.Connect,
+				shardId,
+			};
+
+			const promise = new Promise<void>((resolve) => this.connectPromises.set(shardId, resolve));
+			worker.postMessage(payload);
+			promises.push(promise);
+		}
+
+		await Promise.all(promises);
+	}
+
+	public async destroy(options: Omit<WebSocketShardDestroyOptions, 'recover'> = {}) {
+		const promises = [];
+
+		for (const [shardId, worker] of this.workerByShardId.entries()) {
+			const payload: WorkerSendPayload = {
+				op: WorkerSendPayloadOp.Destroy,
+				shardId,
+				options,
+			};
+
+			const promise = new Promise<void>((resolve) => this.destroyPromises.set(shardId, resolve));
+			worker.postMessage(payload);
+			promises.push(promise);
+
+			await worker.terminate();
```

Comment:
You're killing the worker before awaiting its close, which might mean you don't get to save the session info